android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/PhysicalWebBroadcastService.java

```diff
@@ -110,62 +109,35 @@ public int onStartCommand(Intent intent, int flags, int startId) {
       IntentFilter filter = new IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED);
       registerReceiver(mReceiver, filter);
 
-      Log.d(TAG, mDisplayUrl);
-      byte[] encodedUrl = AdvertiseDataUtils.encodeUri(mDisplayUrl);
-      if (hasValidUrlLength(encodedUrl.length) && checkAndHandleAsciiUrl(mDisplayUrl)) {
-        // Set the url if we can
-        Log.d(TAG, "valid length");
-        mShareUrl = encodedUrl;
-        broadcastUrl();
-      } else {
-        Log.d(TAG, "needs shortening");
-        UrlShortenerClient.ShortenUrlCallback urlSetter =
-            new UrlShortenerClient.ShortenUrlCallback() {
-          @Override
-          public void onUrlShortened(String newUrl) {
-            Log.d(TAG, "shortening success");
-            mShareUrl = AdvertiseDataUtils.encodeUri(newUrl);
-            broadcastUrl();
-          }
-          @Override
-          public void onError(String oldUrl) {
-            Toast.makeText(getApplicationContext(), getString(R.string.shorten_error),
-                Toast.LENGTH_LONG).show();
-            stopSelf();
-          }
-        };
-        UrlShortenerClient shortenerClient = UrlShortenerClient.getInstance(this);
+      handleUrl();
 
-        if (mDisplayUrl.contains("goo.gl")) {
-          // find the site URL and reshorten it since
-          // goo.gl will not reshorten other goo.gl links
-          fetchAndShorten(shortenerClient, urlSetter);
-        } else {
-          shortenerClient.shortenUrl(mDisplayUrl, urlSetter, TAG);
-        }
-      }
       return START_STICKY;
     }
 
-    private void fetchAndShorten(UrlShortenerClient shortenerClient,
-        UrlShortenerClient.ShortenUrlCallback urlSetter) {
-      PwsClient pwsClient = new PwsClient();
-      pwsClient.resolve(Arrays.asList(mDisplayUrl), new PwsResultCallback() {
+    private void handleUrl() {
+        PwsClient pwsClient = new PwsClient();
+        pwsClient.resolve(Arrays.asList(mDisplayUrl), new PwsResultCallback() {
         @Override
         public void onPwsResult(PwsResult pwsResult) {
-          shortenerClient.shortenUrl(pwsResult.getSiteUrl(), urlSetter, TAG);
+            String fullUrl = pwsResult.getSiteUrl();
+            byte[] encodedUrl = AdvertiseDataUtils.encodeUri(fullUrl);
+            if(checkNeedsShortening(encodedUrl, fullUrl)) {
+                shortenAndBroadcastUrl(fullUrl);
+            } else {
+                broadcastUrl(encodedUrl);
+            }
         }
 
         @Override
         public void onPwsResultAbsent(String url) {
-          Toast.makeText(getApplicationContext(), getString(R.string.shorten_error),
+          Toast.makeText(getApplicationContext(), getString(R.string.invalid_url_error),
```

Comment:
We do this pretty often here...could we just make this its own method?  e.g., `toastError(int messageId)`.  You can save this for another change if you want.